pkg/drain/drain.go

```diff
@@ -64,13 +45,33 @@ func GetPodsForDeletion(kc kubernetes.Interface, node string) (pods []v1.Pod, er
 	return pods, nil
 }
 
-// Pared down version of
-// https://github.com/kubernetes/kubernetes/blob/cbbf22a7d2b06a55066b16885a4baaf4ce92d3a4/pkg/kubectl/cmd/drain.go's
-// getDaemonsetController().
-func getDaemonsetController(kc kubernetes.Interface, sr *v1.SerializedReference) (interface{}, error) {
-	switch sr.Reference.Kind {
+// getOwnerDaemonset returns an existing DaemonSet owner if it exists.
+func getOwnerDaemonset(kc kubernetes.Interface, pod v1.Pod) (interface{}, error) {
+	if len(pod.OwnerReferences) == 0 {
+		return nil, fmt.Errorf("pod %q has no owner objects", pod.Name)
+	}
+	for _, ownerRef := range pod.OwnerReferences {
+		// skip pod if it is owned by an existing daemonset
+		if ownerRef.Kind == "DaemonSet" {
+			ds, err := getDaemonsetController(kc, pod.Namespace, &ownerRef)
+			if err == nil {
+				// daemonset owner exists
+				return ds, nil
+			}
+			if !errors.IsNotFound(err) {
+				return nil, fmt.Errorf("failed to get controller of pod %q: %v", pod.Name, err)
+			}
+		}
+	}
+	// pod may have owners, but they don't exist or aren't daemonsets
+	return nil, fmt.Errorf("pod %q has no existing damonset owner", pod.Name)
+}
+
+// Stripped down version of https://github.com/kubernetes/kubernetes/blob/1bc56825a2dff06f29663a024ee339c25e6e6280/pkg/kubectl/cmd/drain.go#L272
+func getDaemonsetController(kc kubernetes.Interface, namespace string, controllerRef *v1meta.OwnerReference) (interface{}, error) {
+	switch controllerRef.Kind {
 	case "DaemonSet":
-		return kc.ExtensionsV1beta1().DaemonSets(sr.Reference.Namespace).Get(sr.Reference.Name, v1meta.GetOptions{})
+		return kc.ExtensionsV1beta1().DaemonSets(namespace).Get(controllerRef.Name, v1meta.GetOptions{})
```

Comment:
still concerned this won't identify daemonsets installed through the apps API group.

Reply:
For what its worth, kubernetes/kubernetes drain (master) only seems to use `client.ExtensionsV1beta1` in detecting DaemonSets.

https://github.com/kubernetes/kubernetes/blob/48f69ac964b9a96b55351a3541f285b4cb5611bb/pkg/kubectl/cmd/drain.go#L341
https://github.com/kubernetes/kubernetes/blob/48f69ac964b9a96b55351a3541f285b4cb5611bb/pkg/kubectl/cmd/drain.go#L362

Reply:
Alright. A typical cluster has a few daemonsets to experiment with:

```
NAMESPACE            NAME                           DESIRED   CURRENT   READY     UP-TO-DATE   AVAILABLE   NODE SELECTOR                     AGE
kube-system          calico-node                    4         4         4         4            4           <none>                            19h
kube-system          kube-apiserver                 1         1         1         1            1           node-role.kubernetes.io/master=   19h
kube-system          kube-proxy                     4         4         4         4            4           <none>                            19h
kube-system          pod-checkpointer               1         1         1         1            1           node-role.kubernetes.io/master=   19h
monitoring           node-exporter                  4         4         4         4            4           <none>                            13h
reboot-coordinator   container-linux-update-agent   4         4         4         4            4           <none>                            13h
```

At present, these daemonsets are all defined as "extensions/v1beta1 DaemonSet" objects. We can

* Switch one to "apps/v1 DaemonSet"
* Intentionally orphan the DaemonSet pods with `kubectl delete daemonset node-exporter --cascade=false`
* Watch `update-agent` logs to see the orphaned pod is deleted ([log line](https://github.com/coreos/container-linux-update-operator/blob/master/pkg/agent/agent.go#L159))

Reply:
Change node-exporter to apps/v1 DaemonSet, it was correctly drained. Change update-agent itself to apps/v1 DaemonSet, it was correctly skipped because it is owned by an existing DaemonSet.

```
I0117 19:48:14.635108       1 agent.go:130] Setting annotations map[string]string{"container-linux-update.v1.coreos.com/reboot-in-progress":"true"}
I0117 19:48:14.655798       1 agent.go:142] Marking node as unschedulable                      
I0117 19:48:14.673086       1 agent.go:147] Getting pod list for deletion                      
I0117 19:48:14.690520       1 agent.go:156] Deleting 1 pods                                    
I0117 19:48:14.690619       1 agent.go:159] Terminating pod "node-exporter-rsgbn"...           
I0117 19:48:14.715594       1 agent.go:171] Waiting for pod "node-exporter-rsgbn" to terminate 
I0117 19:48:24.731064       1 agent.go:374] Deleted pod "node-exporter-rsgbn"                  
I0117 19:48:24.731089       1 agent.go:180] Node drained, rebooting        
```